### PR TITLE
[ie/cbc.ca:player] Fix extractor

### LINKED ISSUE #10170: [cbc.ca:player] Unable to download XML: HTTP Error 404: Not Found
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Non-Geoblocked

### Provide a description that is worded well enough to be understood

I tried to download this video: https://www.cbc.ca/player/play/video/9.6420651

However, it failed and gave me the "Unable to download XML: HTTP Error 404: Not Found" error.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['https://www.cbc.ca/player/play/video/9.6420651', '-o', 'D:\\Downloaded Audio-Video Tracks\\ViaYouTubeDL\\cbc.ca\\Custom\\%(title)s-%(id)s.%(ext)s', '-o', 'D:/EdmontonAirMonitoring.mp4', '-vU']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out cp1252 (No VT), error cp1252 (No VT), screen cp1252 (No VT)
[debug] yt-dlp version stable@2024.05.27 from yt-dlp/yt-dlp [12b248ce6] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.22621-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 6.0-essentials_build-www.gyan.dev (setts), ffprobe 6.0-essentials_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.20.0, brotli-1.1.0, certifi-2024.02.02, curl_cffi-0.5.10, mutagen-1.47.0, requests-2.32.2, sqlite3-3.35.5, urllib3-2.2.1, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets, curl_cffi
[debug] Loaded 1820 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: stable@2024.05.27 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2024.05.27 from yt-dlp/yt-dlp)
[cbc.ca:player] Extracting URL: https://www.cbc.ca/player/play/video/9.6420651
[cbc.ca:player] 9.6420651: Downloading webpage
[ThePlatform] Extracting URL: http://link.theplatform.com/s/ExhSPC/media/guid/2655402169/None?mbr=true&formats=MPEG4,FLV,MP3#__youtubedl_smuggle=%7B%22force_smil_url%22%3A+true%7D
[ThePlatform] None: Downloading SMIL data
[ThePlatform] None: Unable to download XML: HTTP Error 404: Not Found (caused by <HTTPError 404: Not Found>)
  File "yt_dlp\extractor\common.py", line 734, in extract
  File "yt_dlp\extractor\theplatform.py", line 313, in _real_extract
  File "yt_dlp\extractor\theplatform.py", line 34, in _extract_theplatform_smil
  File "yt_dlp\extractor\common.py", line 1133, in download_content
  File "yt_dlp\extractor\common.py", line 1093, in download_handle
  File "yt_dlp\extractor\adobepass.py", line 1366, in _download_webpage_handle
  File "yt_dlp\extractor\common.py", line 954, in _download_webpage_handle
  File "yt_dlp\extractor\common.py", line 903, in _request_webpage
  File "yt_dlp\extractor\common.py", line 890, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 4142, in urlopen
  File "yt_dlp\networking\common.py", line 117, in send
  File "yt_dlp\networking\_helper.py", line 208, in wrapper
  File "yt_dlp\networking\common.py", line 337, in send
  File "yt_dlp\networking\_requests.py", line 366, in _send
yt_dlp.networking.exceptions.HTTPHTTP Error 404: Not Found
An error occured
```

### PR DESCRIPTION
**Pull request by trainman261**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Fixes CBC Player extraction. For now it works with the HLS streaming mechanism that the actual player uses. There is also a way to download the full video in one file, but that is somewhat more complex and I haven't quite gotten behind how that works. If I figure it out, I will add it later.
Two things I'm uncertain about: there are two ID's involved in the videos. There is the so-called "sourceID", which is the ID included in the URL (format: x.xxxxxxx). This ID exists for Videos uploaded using ThePlatform as well as the new system that broke extraction. IDs beginning in 1 seem to be using ThePlatform, IDs beginning in 9 seem to not use it and need this new extraction method, but then there's another ID used to look up the actual video data. I was a little uncertain which ID to use, but I figured that the former ID allows URL matching, so I used that one.
Also, the code to strip the arguments off the thumbnail URL seem to be a little clunky, but I wasn't able to find a more appropriate method in `utils` (I tried `update_url_query` but wasn't able to get that to work).

Fixes #10170 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by taitaitai1777** [comment]
### Title
**Fixing CBC Player Extractor in yt-dlp**

### Summary
To address the broken CBC Player extractor in `yt-dlp`, the solution involves identifying issues through error reproduction, analyzing site changes using developer tools, and modifying extraction logic with updated JSON parsing and URL patterns. Rigorous testing with diverse URLs ensures robustness, followed by documenting changes and conducting code reviews. Finally, a well-documented pull request is submitted, incorporating community feedback for a comprehensive, reliable solution.

### Step-by-Step Guide

1. **Reproduce the Issue:**
   - Install yt-dlp.
     ```sh
     pip install yt-dlp
     ```
   - Attempt downloading a CBC video.
     ```sh
     yt-dlp https://www.cbc.ca/player/play/XXX
     ```
   - Note error message/log.

2. **Analyze Error Messages:**
   - Focus on HTTP errors, JSON parsing issues, or missing keys.

3. **Check CBC Player Changes:**
   - Open CBC video URL in a browser.
   - Use Developer Tools (`F12`) to inspect network requests and DOM structure.

4. **Modify the Extractor Code:**
   - Locate extractor file in `yt-dlp/extractor/`.
   - Update URL patterns and JSON parsing logic.

    ```python
    class CBCPlayerIE(InfoExtractor):
        _VALID_URL = r'https?://www\.cbc\.ca/player/play/[a-z0-9-]+'
        
        def _real_extract(self, url):
            video_id = self._match_id(url)
            webpage = self._download_webpage(url, video_id)
            json_url = self._search_regex(r'"JSON_URL":\s*"([^"]+)"', webpage, 'json url')
            json_data = self._download_json(json_url, video_id)
            video_url = json_data['video']['url']
            title = json_data.get('title', 'Untitled')
            return {
                'id': video_id,
                'title': title,
                'url': video_url,
                'ext': 'mp4',
            }
    ```

5. **Testing:**
   - Test with various CBC URLs.
     ```sh
     yt-dlp https://www.cbc.ca/player/play/XXX
     yt-dlp https://www.cbc.ca/player/play/YYY
     yt-dlp https://www.cbc.ca/player/play/ZZZ
     ```
   - Ensure accurate video and metadata extraction.

6. **Review and Document:**
   - Follow `yt-dlp` coding standards.
   - Update documentation and add comments for clarity.

7. **Submit the Fix:**
   - Create a fork and new branch.
   - Push changes and open a pull request with a detailed description.
     ```
     ### Fix: CBC Player Extractor

     This PR corrects the CBC Player extractor to handle the updated JSON structure and URL patterns on the CBC site.

     **Changes:**

     - Updated JSON URL extraction logic.
     - Adapted JSON parsing to new API responses.
     - Improved regex patterns for URL validation.

     **Testing:**

     - Verified with multiple CBC video URLs.
     - Ensured correct video URL and metadata extraction.

     Closes #1234.
     ```

8. **Community Review:**
   - Engage with feedback.
   - Test across edge cases, refine as needed.

By following these steps, the CBC Player extractor in `yt-dlp` can be effectively fixed and maintained.

**Comment by trainman261** [comment]
@taitaitai1777 not sure I understand. I've followed these steps, or did I miss something?

**Comment by seproDev** [comment]
@trainman261 That comment seems like a ChatGPT generated spam message. You can ignore it.
There are, however, still ruff linting errors. Please address them before someone will have a look at your PR for review.

**Comment by trainman261** [comment]
Ah OK, looking at it a little closer they seem like they might be legitimate complaints this time around. I'll take a look at them hopefully tomorrow and send an update.

**Comment by seproDev** [comment]
We switched from flake8 to ruff. flake8 does not support all the rules we would like to check.

**Comment by trainman261** [comment]
Ah, I hadn't caught that, thanks for the info!

**Comment by trainman261** [comment]
I've fixed one of the complaints from ruff, but the one with the imports has me somewhat confused. I've sorted the imports, and it's still complaining. I haven't been able to find much documentation online as to what could be causing the issue. For some reason I can't get ruff working on my system. Could someone who has a working ruff install please run `ruff --fix`?

File: yt_dlp/extractor/cbc.py [hunk 1]
```diff
@@ -7,16 +7,21 @@
 
 from .common import InfoExtractor
 from ..utils import (
+    base_url,
     ExtractorError,
```
Comment (seproDev) [inline yt_dlp/extractor/cbc.py:11]: Uppercase comes first
```suggestion
    ExtractorError,
    base_url,
```

**Comment by trainman261** [comment]
I've added some code so that the extractor will offer the subtitles as one file, if available. This should probably be done with `traverse_obj`, but I tried back and forth and wasn't able to get it to work. I'll keep playing around and see if I can figure it out, but if someone more knowledgeable can do it quickly, please go ahead.

**Comment by villasv** [comment]
I was looking to download some CBC News videos and stumbled upon this. Thank you @trainman261 for the good work! Hope this gets merged and released soon 🙏

**Comment by nusiloot** [comment]
Using @trainman261 cbc extractor I got:

```
yt-dlp https://www.cbc.ca/player/play/video/9.6459530 --verbose
[debug] Command-line config: ['--ignore-errors', 'https://www.cbc.ca/player/play/video/9.6459530', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2024.07.25 from yt-dlp/yt-dlp [f0993391e] (pip)
[debug] Python 3.10.12 (CPython x86_64 64bit) - Linux-5.15.0-116-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
[debug] Optional libraries: Cryptodome-3.11.0, brotli-1.0.9, certifi-2020.06.20, mutagen-1.45.1, requests-2.32.3, secretstorage-3.3.1, sqlite3-3.37.2, urllib3-2.2.2, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Loaded 1829 extractors
[cbc.ca:player] Extracting URL: https://www.cbc.ca/player/play/video/9.6459530
[cbc.ca:player] 9.6459530: Downloading webpage
[cbc.ca:player] 9.6459530: Downloading JSON metadata
[cbc.ca:player] 9.6459530: Downloading m3u8 information
ERROR: 'NoneType' object has no attribute 'replace'
Traceback (most recent call last):
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1626, in wrapper
    return func(self, *args, **kwargs)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1782, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1841, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 2787, in process_video_result
    self._fill_common_fields(info_dict)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 2708, in _fill_common_fields
    info_dict[old_key] = ', '.join(v.replace(',', '\N{FULLWIDTH COMMA}') for v in new_value)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 2708, in <genexpr>
    info_dict[old_key] = ', '.join(v.replace(',', '\N{FULLWIDTH COMMA}') for v in new_value)
AttributeError: 'NoneType' object has no attribute 'replace'

```

also this error:


```
yt-dlp https://www.cbc.ca/player/play/video/9.6427282 --verbose
[debug] Command-line config: ['--ignore-errors', 'https://www.cbc.ca/player/play/video/9.6427282', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2024.07.25 from yt-dlp/yt-dlp [f0993391e] (pip)
[debug] Python 3.10.12 (CPython x86_64 64bit) - Linux-5.15.0-116-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
[debug] Optional libraries: Cryptodome-3.11.0, brotli-1.0.9, certifi-2020.06.20, mutagen-1.45.1, requests-2.32.3, secretstorage-3.3.1, sqlite3-3.37.2, urllib3-2.2.2, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Loaded 1829 extractors
[cbc.ca:player] Extracting URL: https://www.cbc.ca/player/play/video/9.6427282
[cbc.ca:player] 9.6427282: Downloading webpage
[cbc.ca:player] 9.6427282: Downloading JSON metadata
ERROR: Unable to handle request: Unsupported url scheme: "" (requests, urllib, websockets)
Traceback (most recent call last):
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1626, in wrapper
    return func(self, *args, **kwargs)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1761, in __extract_info
    ie_result = ie.extract(url)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 740, in extract
    ie_result = self._real_extract(url)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/extractor/cbc.py", line 325, in _real_extract
    self._download_json(info['media']['assets'][0]['key'], video_id)['url'], video_id)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 1139, in download_content
    res = getattr(self, download_handle.__name__)(url_or_request, video_id, **kwargs)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 1099, in download_handle
    res = self._download_webpage_handle(
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 960, in _download_webpage_handle
    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data,
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 896, in _request_webpage
    return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query, extensions))
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 4165, in urlopen
    return self._request_director.send(req)
  File "/home/aby/.local/lib/python3.10/site-packages/yt_dlp/networking/common.py", line 130, in send
    raise NoSupportingHandlers(unsupported_errors, unexpected_errors)
yt_dlp.networking.exceptions.NoSupportingHandlers: Unable to handle request: Unsupported url scheme: "" (requests, urllib, websockets)

```

Do I miss something?

**Comment by trainman261** [comment]
Things seem to be in flux on the CBC Player at the moment. I've also noticed another issue that has cropped up in the tests. I hope I'll be able to get around to it tomorrow.
CBC seems to be moving entirely off of ThePlatform, and doing it bit by bit. Every bit seems to break something here or there, but studying what changes have happened so far, I hope the last bit that has changed is the last step in the migration process, so that things will be a little more stable moving forward.

**Review by bashonly — APPROVED** [review]
I started looking into the issues brought up in the comments here, and I ended up taking what you had done and refactoring/fixing/finishing it. It looks good to me now. Test it out and let me know what you think

File: yt_dlp/extractor/cbc.py [hunk 2]
```diff
@@ -307,23 +373,85 @@ class CBCPlayerIE(InfoExtractor):
...
-                f'http://link.theplatform.com/s/ExhSPC/media/guid/2655402169/{video_id}?mbr=true&formats=MPEG4,FLV,MP3', {
-                    'force_smil_url': True,
-                }),
+            **traverse_obj(data, {
+                'title': ('title', {str}),
+                'description': ('description', {str.strip}),
+                'thumbnail': ('image', 'url', {url_or_none}, {functools.partial(update_url, query=None)}),
+                'timestamp': ('publishedAt', {functools.partial(float_or_none, scale=1000)}),
+                'media_type': ('media', 'clipType', {str}),
+                'is_live': ('streamType', {lambda x: x == 'Live'}),
```
Comment (bashonly) [inline yt_dlp/extractor/cbc.py:442]:
```suggestion
                'is_live': ('media', 'streamType', {lambda x: x == 'Live'}),
```

**Comment by trainman261** [comment]
I've looked over it, for the most part it works, with one exception: it extracts both the subtitles via HLS as well as as a single file, however, it prefers the subtitles via HLS. Problem is, this is terrible for performance, as it has to fetch sometimes hundreds of tiny file instead of one pretty small file. The lines:
`subtitles.setdefault(sub.get('language') or 'und', []).append({
   'url': sub['src'],
   'name': sub.get('label'),
})`
would work _if_ CBC used the consistent language codes, but for whatever reason, they don't. For example in the test of The National (the full show): the subtitle file has `en-US`, and the m3u8 format has `eng`. That was the reason I only included the m3u8 format if it didn't find the file.
Maybe there's some way to mark subtitle entries as having a higher priority that I missed?

I've also figured out how to get the full video file - I have implemented it, but I haven't gotten the tests working yet, since for some reason not being able to find the full video file results in a failure, even though I've passed `fatal=False`. I've committed it for now, if you would rather get this merged first and take care of that later, I can always take back the commit. Otherwise I hope to figure it out on Tuesday.

**Comment by bashonly** [comment]
Please do not force push

**Comment by trainman261** [comment]
> Please do not force push

Sorry, I just took back a commit of my own that messed something up, nothing else.

**Comment by villasv** [comment]
I have been using the branch directly and it has been working for me so far 🚀

**Comment by bashonly** [comment]
> Sorry, I just took back a commit of my own that messed something up, nothing else.

@trainman261 All good. Just a note for the future. Thanks for your work on this one

**Comment by trainman261** [comment]
No problem, glad I could help - and thanks for fixing it up! It now works how I would expect it to on my end.
